# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Linear interpolation of genetic map values
NULL

#' @export
interpolate_genetic_map <- function(map, map_pos, target_pos, strict = TRUE, progress = FALSE) {
    .Call('_ldmap_interpolate_genetic_map', PACKAGE = 'ldmap', map, map_pos, target_pos, strict, progress)
}

rcppeigen_hello_world <- function() {
    .Call('_ldmap_rcppeigen_hello_world', PACKAGE = 'ldmap')
}

rcppeigen_outerproduct <- function(x) {
    .Call('_ldmap_rcppeigen_outerproduct', PACKAGE = 'ldmap', x)
}

rcppeigen_innerproduct <- function(x) {
    .Call('_ldmap_rcppeigen_innerproduct', PACKAGE = 'ldmap', x)
}

rcppeigen_bothproducts <- function(x) {
    .Call('_ldmap_rcppeigen_bothproducts', PACKAGE = 'ldmap', x)
}

sorted_snp_df <- function(chr, pos) {
    .Call('_ldmap_sorted_snp_df', PACKAGE = 'ldmap', chr, pos)
}

set_ld_region <- function(ld_chr, ld_start, ld_stop, ld_region_id, chr, pos, assign_all = TRUE) {
    .Call('_ldmap_set_ld_region', PACKAGE = 'ldmap', ld_chr, ld_start, ld_stop, ld_region_id, chr, pos, assign_all)
}

#' Determine whether 2 alleles are compatible
NULL

#' @export
strand_flip <- function(ref_alt, reverse = FALSE) {
    .Call('_ldmap_strand_flip', PACKAGE = 'ldmap', ref_alt, reverse)
}

#' Find query SNP in a list of reference snps
#' @param query_chrom query chromosome
#' @param query_pos query position
#' @param ref_chrom reference chromosome
#' @param ref_pos reference position
#' @param query_chunk region assignment for query (optional)
#' @param ref_chunk region assignment for reference (optional)
#' @return returns a vector with the position of the match, or NA if no match is found.
#' @export
find_alleles <- function(query_chrom, query_pos, ref_chrom, ref_pos, query_chunk = as.integer( c()), ref_chunk = as.integer( c())) {
    .Call('_ldmap_find_alleles', PACKAGE = 'ldmap', query_chrom, query_pos, ref_chrom, ref_pos, query_chunk, ref_chunk)
}

#' Determine whether 2
#' @param query_ref_alt is a ref/alt pair
#' @param target_ref_alt is another ref/alt pair
#' @return returns a vector with 1 if the query matches the target, -1 if a flip is required, or 0 if they are incompatible;
#' Note that in the case that the reference and/or alternate allele are not single characters, they will only be checked for equality
#' @export
flip_alleles <- function(query_ref_alt, target_ref_alt) {
    .Call('_ldmap_flip_alleles', PACKAGE = 'ldmap', query_ref_alt, target_ref_alt)
}

